### DISCUSSION TITLE
deviceConfigPriorityDir not working

### DISCUSSION THREAD

**Discussion by glyph-se** [zwave-js/zwave-js#3063] [General] [answered]
I'm not sure if I encountered a bug or if I am doing something wrong. I am still very new to fiddling with device configuration files.

I'm running Home Assistant OS with the Z-Wave JS addon.

What I have done is:

* Install the SSH & Web Terminal HA addon in non-protected mode
* Start a shell inside the `addon_core_zwave_js` container
* Modify the `/etc/zwave_config.json` file by adding `"deviceConfigPriorityDir": "/data/cache/config",` to it.
* Copy one of the existing files `cp /data/db/devices/0x010f/fgs223.json .` to the above directory.
* Also copy over the `master_template.json` file.
* Restart zwave-js server with `kill <pid>`
* Looking at the zwave-js logs by `docker logs addon_core_zwave_js` shows me

```
...
2021-07-21T11:19:39.263Z CNTRLR   [Node 008] Embedded device config loaded
2021-07-21T11:19:39.278Z CONFIG   Error loading device config /data/cache/config/fibaro/fgs223.json
2021-07-21T11:19:39.280Z CNTRLR   [Node 009] No device config found
2021-07-21T11:19:39.334Z CNTRLR   [Node 010] Embedded device config loaded
...
```

The `fgs223.json` is an exact copy of the default file. What causes this error? It also appears that when parsing it fails z-wave js proceeds to load the built in file without saying so in the logs. I have tried increasing the log level to debug but that doesn't show me anything more related to the device config.

My `config` folder looks like this

```
bash-5.1# find /data/cache/config/
/data/cache/config/
/data/cache/config/templates
/data/cache/config/templates/master_template.json
/data/cache/config/fibaro
/data/cache/config/fibaro/fgs223.json
```

Without the `master_template.json` file I got errors saying that the template wasn't found, so I guess something is working. At first I suspected there were errors in my modified device config file, thus I tried with the default file for fgs223. Did I miss something in my steps? Any way to get more details about the error and why loading the device configuration file fails?

I guess you might need more info to debug this or help me out, but I don't know what that would be.

**Comment by AlCalzone** [comment, marked as answer]
Since you are already fiddling around in the files, can you edit `node_modules/@zwave-js/config/build/ConfigManager.js`?
There should be a method called `lookupDevicePreserveConditions`, in which you'll find a line like this (probably around line 382, give or take):
```js
this.logger.print(`Error loading device config ${filePath}`, "error");
```

Change that to 
```js
this.logger.print(`Error loading device config ${filePath}: ${e}`, "error");
```
save and restart.

Then we should at least know why the loading fails.

**Reply by glyph-se** [reply]
Thanks for your quick response, the exception message is:

```
2021-07-21T12:41:39.231Z CONFIG   Error loading device config /data/cache/config/fibaro/fgs223.json: ZWaveError:
                                   Tried to import config file "/data/cache/config/fibaro/fgs223.json" outside o
                                  f root directory "/usr/src/node_modules/@zwave-js/config/config/devices"!
```

Do I need to change some other configuration to allow this?

Edit: Could this be a bug? This validation appears to have been introduced recently in https://github.com/zwave-js/node-zwave-js/pull/3035/files#diff-e311357f76159154dffc6b9328647b48fbc3af6282c9f2a8c0b62adb114eb640

But I'm no JavaScript / TypeScript developer and I don't know much about this project, so just guessing.

**Reply by AlCalzone** [reply]
Yeah, looks like it doesn't correctly use the priorityDir for validation

**Reply by AlCalzone** [reply]
==> https://github.com/zwave-js/node-zwave-js/issues/3065

**Reply by glyph-se** [reply]
Thanks for the quick fix. I no longer get the error in the logs and it now correctly shows the line

```
2021-07-23T08:29:04.293Z CNTRLR   [Node 009] User-provided device config loaded
```

Though I am still unable to get the modified configuration file to show up in the Home Assistant UI.

I'm still using the original file with just one modification to a label to test this. If I modify the file directly in `/data/db/devices/0x010f/fgs223.json` the modification shows up after a re-interview. This is despite there being a file in the deviceConfigPriorityDir and it showing a user-provided file being loaded in the logs.

Currently using zwave-js version 8.0.5.

I've checked the debug logs without any clues to what to try.

The logs are available here https://dflund.se/~glyph/stuff/zwave-js/2021-07-23-user-provided-config-file-fails.txt , last interview is between 2021-07-23T08:28:52.992Z and 2021-07-23T08:29:10.832Z

I have been trying to check different documentation and forums for pointers what I might be missing but I haven't found much on this subject.

Would be happy for some help as what to try next. Could this be caused by the different ways zwave-js is used by zwavejs2mqtt vs core_zwave_js addon?

**Reply by AlCalzone** [reply]
That config file stuff is making me gray these last few days. Apparently, 8.0.5 considers the external location for the embedded config as "user-provided", so that log line doesn't tell us anything sadly. It's already fixed on master, but not yet released.
